streams/src/main/java/org/apache/kafka/streams/state/QueryableStoreTypes.java

```diff
@@ -0,0 +1,89 @@
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.kafka.streams.state;
+
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.state.internals.CompositeReadOnlyStore;
+import org.apache.kafka.streams.state.internals.CompositeReadOnlyWindowStore;
+
+/**
+ * Provides access to the {@link QueryableStoreType}s provided with KafkaStreams. These
+ * can be used with {@link org.apache.kafka.streams.KafkaStreams#getStore(String, QueryableStoreType)}
+ * To access and query the {@link StateStore}s that are part of a Topology
+ */
+public class QueryableStoreTypes {
+
+    /**
+     * A {@link QueryableStoreType} that accepts {@link ReadOnlyKeyValueStore}
+     * @param <K>   key type of the store
+     * @param <V>   value type of the store
+     * @return  {@link KeyValueStoreType}
+     */
+    public static <K, V> QueryableStoreType<ReadOnlyKeyValueStore<K, V>> keyValueStore() {
+        return new KeyValueStoreType<>();
+    }
+
+    /**
+     * A {@link QueryableStoreType} that accepts {@link ReadOnlyWindowStore}
+     * @param <K>   key type of the store
+     * @param <V>   value type of the store
+     * @return  {@link WindowStoreType}
+     */
+    public static <K, V> QueryableStoreType<ReadOnlyWindowStore<K, V>> windowStore() {
+        return new WindowStoreType<>();
+    }
+
+    public static abstract class QueryableStoreTypeMatcher<T> implements QueryableStoreType<T> {
+
+        private final Class matchTo;
+
+        public QueryableStoreTypeMatcher(Class matchTo) {
+            this.matchTo = matchTo;
+        }
+        @SuppressWarnings("unchecked")
+        @Override
+        public boolean accepts(final StateStore stateStore) {
+            return matchTo.isAssignableFrom(stateStore.getClass());
+        }
+    }
+
+    private static class KeyValueStoreType<K, V> extends
+                                                 QueryableStoreTypeMatcher<ReadOnlyKeyValueStore<K, V>> {
+        KeyValueStoreType() {
+            super(ReadOnlyKeyValueStore.class);
+        }
+
+        @Override
+        public ReadOnlyKeyValueStore<K, V> create(final StateStoreProvider storeProvider,
+                                                  final String storeName) {
+            return new CompositeReadOnlyStore<>(storeProvider, this, storeName);
+        }
+
+    }
+
+    private static class WindowStoreType<K, V> extends QueryableStoreTypeMatcher<ReadOnlyWindowStore<K,
+            V>> {
+        WindowStoreType() {
+            super(ReadOnlyWindowStore.class);
```

Comment:
We want `ReadOnlyWindowStore.class` as it restricts the API to only `fetch`. `WindowStore extends ReadOnlyWindowStore`